### ISSUE TITLE
fromDataUnion missing default return path?

### ISSUE THREAD

**Issue by croser** [vojtechhabarta/typescript-generator#376]
hi,
when using @JsonTypeInfo and @JsonSubTypes code as the following is generated:

```
static fromDataUnion(data: ChangeUnion): ChangeUnion {
        if (!data) {
            return data;
        }
        switch (data.changeType) {
            case "OBJECT":
                return ObjectChange.fromData(data);
            case "OTHER":
                return OtherChange.fromData(data);
        }
    }
```

The typescript compiler complains with `TS2366: Function lacks ending return statement and return type does not include undefined`.

Thanks!

**Comment by vojtechhabarta** [comment]
hi, how does `ChangeUnion` type look like? Does it have two types?

**Comment by croser** [comment]
exactly:

```
export type ChangeUnion = ObjectChange | OtherChange;
```

**Comment by vojtechhabarta** [comment]
I still don't know what might be the problem.
Could it be the same as #375?
How does `changeType` property look like in `ObjectChange` and `OtherChange` types?

**Comment by croser** [comment]
hi, the problem is that the generated function does not have a default return value. no default in the switch and no return if changeType is neither "OBJECT" nor "OTHER". but a fallback return is required, as the method expects a ChangeUnion to be returned. Easiest solution would be to insert `return data;` as the last line.

**Comment by vojtechhabarta** [comment]
Yes, but purpose of this method is to tranform all possible inputs. TypeScript compiler should know that we are handling all possible cases in the `switch` statement and that this method cannot return `undefined` by default. What value can `data.changeType` have other than `"OBJECT"` and `"OTHER"`?

In other words I think `return data;` just hides problem that is elsewhere.

**Comment by vojtechhabarta** [comment]
Could you please paste here `Change`, `ObjectChange` and `OtherChange` Java classes? Declaration with annotations should be sufficient. I would like to reproduce the problem.

**Comment by croser** [comment]
hi,

in the meantime i found out that i can fix this by adding the following line after the switch in `cz.habarta.typescript.generator.ext.JsonDeserializationExtension.createDeserializationMethodForTaggedUnion(SymbolTable, TsModel, TsBeanModel)`:

```
body.add(new TsReturnStatement(new TsIdentifierReference("data")));
```

for reference and further analysis, see the annotated java code below. thanks!

```
// Change.java

@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "changeType", visible = true)
@JsonSubTypes({
  @JsonSubTypes.Type(value = ObjectChange.class, name = "OBJECT"),
  @JsonSubTypes.Type(value = OtherChange.class, name = "OTHER"),
})
public class Change {
  @JsonProperty("changeType")
  private String changeType;

  public String getChangeType() {
    return changeType;
  }

  public void setChangeType(String changeType) {
    this.changeType = changeType;
  }
}

// ObjectChange.java

public class ObjectChange extends Change {
  @JsonProperty("changeType")
  private String changeType = "OBJECT";

  public String getChangeType() {
    return changeType;
  }

  public void setChangeType(String changeType) {
    this.changeType = changeType;
  }
}

// OtherChange.java

public class OtherChange extends Change {
  @JsonProperty("changeType")
  private String changeType = "OTHER";

  @JsonProperty("description")
  private String description;

  public String getChangeType() {
    return changeType;
  }

  public void setChangeType(String changeType) {
    this.changeType = changeType;
  }
  
  public String getDescription() {
    return description;
  }

  public void setDescription(String description) {
    this.description = description;
  }
}
```

**Comment by vojtechhabarta** [comment]
Root cause of this problem is `changeType` property. Jackson duplicates this property in generated JSON, see #392. I would recommend to remove this property from the classes.

In [v2.16.538](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.16.538) I changed how issue with duplicate property is handled. So if you still want to retain `changeType` property typescript-generator will not generate tagged union and generated code should be valid.

**Comment by croser** [comment]
Thanks for that! 

Well, removing this property is not so simple, as this code is being generated by an openapi generator. The correct way might be to use EXISTING_PROPERTY instead of PROPERTY, which is not supported by your generator: See #390.

**Comment by vojtechhabarta** [comment]
In #582 I added support for `EXISTING_PROPERTY`, it is released in [v2.28.785](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.28.785).
